core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java

```diff
@@ -0,0 +1,233 @@
+      final Map<Address, InboundTransferTask> transferTaskMap = new HashMap<>();
+      volatile CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> future;
+
+      SegmentRequest(int segmentId, LocalizedCacheTopology topology) {
+         this.segmentId = segmentId;
+         this.topology = topology;
+         this.replicaHosts = topology.getDistributionForSegment(segmentId).writeOwners();
+      }
+
+      synchronized CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> requestState() {
+         assert future == null;
+         if (trace) log.tracef("Attempting to receive replicas for segment %s from %s with topology %s", segmentId, replicaHosts, topology);
+
+         List<CompletableFuture<Void>> completableFutures = new ArrayList<>();
+         for (Address replica : replicaHosts) {
+            if (replica.equals(rpcManager.getAddress())) {
+               dataContainer.forEach(entry -> {
+                  int keySegment = topology.getDistribution(entry.getKey()).segmentId();
+                  if (keySegment == segmentId) {
+                     addKeyToReplicaMap(replica, entry);
+                  }
+               });
+            } else {
+               InboundTransferTask transferTask = createTransferTask(segmentId, replica, topology);
+               transferTaskMap.put(replica, transferTask);
+               completableFutures.add(transferTask.requestSegments());
+            }
+         }
+         future = CompletableFuture
+               .allOf(completableFutures.toArray(new CompletableFuture[completableFutures.size()]))
+               .thenApply(aVoid -> {
+                  List<Map<Address, InternalCacheEntry<K, V>>> retVal = keyReplicaMap.entrySet().stream()
+                        .map(Map.Entry::getValue)
+                        .collect(Collectors.toList());
+                  keyReplicaMap.clear();
+                  transferTaskMap.clear();
+                  requestMap.remove(segmentId);
+                  return Collections.unmodifiableList(retVal);
+               });
+
+         // If an exception is thrown by any of the inboundTransferTasks, then remove all segment results and cancel all tasks
+         future.exceptionally(throwable -> {
+            if (trace) log.tracef(throwable, "Exception when processing InboundTransferTask for cache %s", cacheName);
+            cancel(throwable);
+            return null;
+         });
+         return future;
+      }
+
+      synchronized void receiveState(Address sender, int topologyId, Collection<StateChunk> stateChunks) {
+         if (topologyId < topology.getTopologyId()) {
+            if (trace)
+               log.tracef("Discarding state response with old topology id %d for cache %s, the smallest allowed topology id is %d",
+                     topologyId, topology.getTopologyId(), cacheName);
+            return;
+         }
+
+         InboundTransferTask transferTask = transferTaskMap.remove(sender);
```

Comment:
remove? what if this isn't the last chunk? the next chunk wouldn't find the task